text/0000-Bundle-Load-API.md

```diff
@@ -0,0 +1,136 @@
+## Summary
+
+Currently there is no API function in Bundle class to load bundles/dynamic libraries and retrieve desired symbols at runtime.
+This document discusses need for such API in the particular use case of declarative services (referred as DS henceforth in this document) and its possible design/solution.
+
+## Motivation
+
+Developers using DS rely on DS's SCR (service component runtime) to delay/lazily load the bundle's shared libraries. 
+These bundles are loaded by calling extern "C" functions which internally call the native OS APIs.
+
+One could argue that Bundle class's Start() function could be used OR modified to accomodate this functionality instead of adding a new member function in the Bundle class.
+However, this is not possible because DS runtime can't call Bundle's Start() function as DS is waiting on the BUNDLE_STARTED event. This event is fired when the bundle is activated.
+On a related note, since the DS bundles are marked as  "bundle.activator : false" , Start() function can not load the bundle's symbols from its shared library.
+
+Given this, DS can't use bundle's Start() function to load the shared library associated with the bundle and needs a way to access the shared library's interface.
+As such, DS uses OS specific API "load library" calls for retrieving bundles/shared objects and resolving symbols.
+
+This loading of libraries/bundles increases the overhead in the declarative services while doing the symbol resoution and ref count increment done in the system.
+
+Adding to that, this functionality to load bundles/shared objects is currently implemented at two places
+
+1) Declarative services runtime
+2) framework 
+
+During the application run, these different callsites may become out of sync as CppMicroServices and DS are maintained.
+
+Given this it seems appropriate to have a new Bundle::LoadSymbols API for CppMicroServices library.
+
+## Requirements
+
+For the specific use case of DS (and similar extender clients), have an API in the CppMicroServices that
+
+1) loads shared library without dealing with OS specific data types
+2) resolves symbols in a Bundle's shared library with the intent of calling exported functions
+3) does not require the users of this API to write any OS specific code e.g DS would only call bundle.GetSymbol() instead of using native OS calls
+
+## Detailed design
+
+Bundle::GetSymbols() API will be declared in the file CppMicroServices/framework/include/cppmicroservices/Bundle.h as below
+The output of above API will be in the wrapper class which will be as belows.
+
+```
+/**
+* @brief  This class holds the output of GetSymbol API.
+*
+* It acts as a wrapper over the handles returned by the OS APIs that loads the library.
+* Clients have to supply the function type as template argument to correctly get the desired symbol resolved using
+* GetSymbols function. See the usage workflow for more details.
+*
+*/
+
+namespace cppmicroservices {
+ls ..libmwsplashscreen.so
+class BundleHandle
```

Comment:
Is the ```BundleHandle``` class necessary? It seems that there could be one function, ```Bundle::GetSymbols``` that will do what is required to retrieve a function pointer from the shared library associated with the bundle object and return it to the caller.

Reply:
Hi Jeff, 

This sounds better than keeping an extra class for storing the shared lib handles and then using them to retrieve the symbols.
I will make the change.

Thanks,
AD